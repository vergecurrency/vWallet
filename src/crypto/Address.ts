export interface Address {
  version: string
  createdOn: number
  address: string
  walletId: string
  isChange: boolean
  path: string
  publicKeys: string[]
  coin: string
  network: string
  type: string
  _id: string
}
