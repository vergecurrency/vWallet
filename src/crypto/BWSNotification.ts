export interface Address {
  address: string
  result: string
  error: string
}

export interface Sync {
  result: string
  error: string
}

export interface BWSNotification {
  version: string
  createdOn: number
  id: number
  type: string
  data: Address | Sync
  walletId: string
  creatorId: string
}
