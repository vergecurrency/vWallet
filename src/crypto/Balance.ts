export interface Balance {
  /**
   * General (unconfirmed/cofirmed) balances
   */
  totalAmount: number
  availableAmount: number
  lockedAmount: number

  /**
   * General only confirmed balances
   */
  totalConfirmedAmount: number
  availableConfirmedAmount: number
  lockedConfirmedAmount: number

  /**
   * Addresses that hold any value
   */
  byAddress: any[]
}
