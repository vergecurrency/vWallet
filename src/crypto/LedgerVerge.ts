import { withDevice } from './helpers/deviceAccess'
import btc from './helpers/getAddressForCurrency/btc'
import hwAppBtc from '@ledgerhq/hw-app-btc'

const coinDefinition = {
  id: 'verge',
  coinType: 1,
  name: 'Verge',
  managerAppName: 'Verge',
  ticker: 'XVG',
  scheme: 'verge',
  color: '#ffae35',
  units: [
    {
      name: 'bitcoin',
      code: 'BTC',
      symbol: 'Ƀ',
      magnitude: 8,
    },
    {
      name: 'mBTC',
      code: 'mBTC',
      symbol: 'Ƀ',
      magnitude: 5,
    },
    {
      name: 'bit',
      code: 'bit',
      symbol: 'Ƀ',
      magnitude: 2,
    },
    {
      name: 'satoshi',
      code: 'sat',
      symbol: 'Ƀ',
      magnitude: 0,
    },
  ],
  supportsSegwit: false,
  family: 'bitcoin',
  ledgerExplorerId: 'xvg',
  blockAvgTime: 0.5 * 60,
  bitcoinLikeInfo: {
    P2PKH: 30,
    P2SH: 33,
  },
}

export enum sigHashType {
  SIGHASH_ALL = 1,
  SIGHASH_NONE = 2,
  SIGHASH_SINGLE = 3,
  SIGHASH_FORKID = 0x40,
  SIGHASH_ANYONECANPAY = 0x80
}

export const getPublicWalletAddress = async ({
  devicePath,
  path,
  ...options
}) =>
  withDevice(devicePath)(transport => {
    return btc(transport, coinDefinition, path, options)
  }).catch(e => {
    if (e && e.name === 'TransportStatusError') {
      if (e.statusCode === 0x6b00 && options.verify) {
        throw new Error('0x6b00')
      }
      if (e.statusCode === 0x6985) {
        throw new Error('0x6985')
      }
    }
    throw e
  })

export function encode({ type, version, xpub, walletName }) {
  return `${type}:${version}:${xpub}:${walletName}`
}

export function decode(accountId: string): any {
  const splitted = accountId.split(':')
  const [type, version, xpub, walletName] = splitted
  return { type, version, xpub, walletName }
}

export function bigNumberToLibcoreAmount(core: any, njsWalletCurrency: *, bigNumber: ) {
  return new core.NJSAmount(njsWalletCurrency, 0).fromHex(njsWalletCurrency, bigNumber.toString(16))
}

export const buildNewTransaction = async (walletName, index, core, normalAmount, fee, recipient) => {
  const njsWallet = await core.getPoolInstance().getWallet(walletName)
  const njsAccount = await njsWallet.getAccount(index)
  const bitcoinLikeAccount = njsAccount.asBitcoinLikeAccount()
  const njsWalletCurrency = njsWallet.getCurrency()
  const amount = bigNumberToLibcoreAmount(
    core,
    njsWalletCurrency,
    normalAmount,
  )
  const fees = bigNumberToLibcoreAmount(
    core,
    njsWalletCurrency,
    fee,
  )
  const transactionBuilder = bitcoinLikeAccount.buildTransaction()

  // TODO: check if is valid address. if not, it will fail silently on invalid

  transactionBuilder.sendToAddress(amount, recipient)
  // TODO: don't use hardcoded value for sequence (and first also maybe)
  transactionBuilder.pickInputs(0, 0xffffff)
  transactionBuilder.setFeesPerByte(fees)

  return await transactionBuilder.build()
}

export const signVergeTransaction = async (
  devicePath,
  currencyId,
  builtTransaction,
  sigHashType,
) =>
  withDevice(devicePath)(async transport =>
    signTransaction({
      currencyId,
      hwApp: new hwAppBtc(transport),
      transaction: builtTransaction,
      sigHashType: parseInt(sigHashType, 16),
      supportsSegwit: false,
      isSegwit: false,
      hasTimestamp: true,
    }),
  )
