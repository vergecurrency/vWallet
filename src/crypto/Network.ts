export default {
  verge: {
    bip32: {
      public: 0x0488b21e,
      private: 0x0488ade4,
    },
    pubKeyHash: 0x1e,
    scriptHash: 0x33,
    wif: 0x80,
  },
}
