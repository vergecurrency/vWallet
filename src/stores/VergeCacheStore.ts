let vergeCacheStore
if (process.env.NODE_ENV !== 'test') {
  const electronStore = require('electron-store')
  vergeCacheStore = new electronStore({
    encryptionKey: Buffer.from('vergecurrency'),
  })
} else {
  let testStore = {}
  vergeCacheStore = {
    get: name => testStore[name],
    set: (name, data) => {
      testStore = { ...testStore, [name]: data }
    },
    clean: () => {
      testStore = {}
    },
  }
}

export default vergeCacheStore
