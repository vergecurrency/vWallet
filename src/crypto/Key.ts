// nodejs module - https://nodejs.org/api/crypto.html
import { createECDH, randomBytes, ECDH, createHash } from 'crypto'
import * as Base58 from 'bs58'
// create curve sepcification based on our coin.
const ecdh: ECDH = createECDH('secp256k1')

export const generateRandomSeed = (size: number = 32): Buffer =>
  randomBytes(size)

export const bufferToHexRepresentation = (bufferInput: Buffer): string =>
  Buffer.from(bufferInput).toString('hex')

export const getPublicKeyFromPrivateKey = (
  privateKey: Buffer | string,
  formatType: 'compressed' | 'uncompressed',
): Buffer => {
  if (privateKey instanceof Buffer) {
    ecdh.setPrivateKey(privateKey)
  } else if (typeof privateKey === 'string') {
    ecdh.setPrivateKey(privateKey, 'hex')
  } else {
    throw Error('Cannot parse your private key type.')
  }

  return Buffer.from(ecdh.getPublicKey('hex', formatType))
}

export const getPublicKeyHash = (publicKey: Buffer): Buffer => {
  return createHash('rmd160')
    .update(
      createHash('sha256')
        .update(publicKey)
        .digest(),
    )
    .digest()
}

export const doubleSHA256 = (secret: Buffer): Buffer =>
  createHash('sha256')
    .update(
      createHash('sha256')
        .update(secret)
        .digest(),
    )
    .digest()

export const createAddress = (payload: Buffer, version: number): string => {
  const buffer = Buffer.concat([Buffer.from([version]), payload])
  const checksum = doubleSHA256(buffer).slice(0, 4)

  return Base58.encode(Buffer.concat([buffer, checksum]))
}
