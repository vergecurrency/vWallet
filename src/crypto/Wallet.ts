import * as Client from '../../bitcore/packages/bitcore-wallet-client/'
import settings from '../settings'
import { randomBytes } from 'crypto'
import { Balance } from './Balance'
import { Address } from './Address'
import { WalletTransaction } from './Transaction'
import { OutgoingTransaction } from './OutgoingTransaction'
import VergeCacheStore from '../stores/VergeCacheStore'
import { BWSNotification } from './BWSNotification'
import { logger } from '../utils/Logger'
import { Constants } from '../Constanst'
const invariant = require('invariant')

const client = new Client({
  baseUrl: settings.BITPAY_WALLET.BWS_INSTANCE_URL,
  verbose: settings.BITPAY_WALLET.VERBOSE,
  doNotVerifyPayPro: true,
})

interface CreatedWalletData {
  mnemonic: string
}

export class VergeLightClient {
  private readonly client: any
  private readonly KEY: string = 'WALLET'
  private password: string
  private scannedErrors: number[] = []
  private isScanning: boolean = true

  constructor(client) {
    this.client = client

    if (this.isWalletAlreadyExistent()) {
      this.client.import(VergeCacheStore.get(this.KEY))
    }
  }

  /**
   * Checks if the wallet that is being used
   * is ready for full usage!
   *
   * (best check for start up a wallet)
   */
  public isWalletReady(): boolean {
    return Boolean(this.client.isComplete())
  }

  /**
   * Checks if the currently used wallet is encrypted
   *
   * Kinda like a locked/unlocked check
   */
  public isWalletLocked(): boolean {
    return Boolean(this.client.isPrivKeyEncrypted())
  }

  public isWalletAlreadyExistent(): boolean {
    return typeof VergeCacheStore.get(this.KEY) === 'string'
  }

  /**
   * Creates OR overrides a wallet
   * @param passphrase
   */
  public createNewWallet(passphrase: string): Promise<CreatedWalletData> {
    invariant(!!passphrase, 'No Empty passphrase allowed')
    // check that we aren't mindlessly overwriting wallets
    invariant(
      !this.isWalletAlreadyExistent(),
      'Existing wallet can`t be overriden!',
    )
    // creates a new wallet with all data (privKey,Mnemonic ...)
    this.client.seedFromRandomWithMnemonic({
      passphrase,
      coin: settings.COIN,
      network: settings.NETWORK,
      account: settings.DEFAULT_ACCOUNT, // account zero, we won't support multiple acc
    })

    return new Promise(resolve => {
      this.client.createWallet(
        'wallet',
        // removing names because of tracking
        randomBytes(128).toString('hex'),
        1, // N
        1, // M
        { network: settings.NETWORK, coin: settings.COIN },
        err => {
          // check if we received errors
          invariant(!err, `Error happend while creating wallet: ${err}`)

          this.exportEncryptedWalletToFile(passphrase)

          return resolve({
            mnemonic: this.client.credentials.mnemonic,
          })
        },
      )
    })
  }

  exportEncryptedWalletToFile(passphrase: string) {
    // when encryption is off for now, then keep state the same after wards
    if (!this.client.isPrivKeyEncrypted()) {
      this.client.encryptPrivateKey(passphrase)
      VergeCacheStore.set(this.KEY, this.client.export())
      this.client.decryptPrivateKey(passphrase)
    } else {
      // otherwise just write it and keep it locked.
      VergeCacheStore.set(this.KEY, this.client.export())
    }
  }

  async unlock(passphrase: string): Promise<boolean> {
    try {
      // when encryption is off for now, then keep state the same after wards
      if (this.client.isPrivKeyEncrypted()) {
        this.client.decryptPrivateKey(passphrase)
        this.password = passphrase
        return !this.client.isPrivKeyEncrypted()
      }
    } catch (e) {
      return false
    }

    return true
  }

  lock() {
    if (!this.client.isPrivKeyEncrypted()) {
      this.client.encryptPrivateKey(this.password)
      delete this.password
    }
  }

  public getBalance(): Promise<Balance> {
    return new Promise((resolve, reject) => {
      if (this.isScanning) {
        reject('Scanning wallet.')
      }
      // no options needed for our wallet
      client.getBalance({}, (err, balance: Balance) => {
        if (err) reject(err)

        resolve(balance)
      })
    })
  }

  public getAddress(): Promise<Address> {
    return new Promise((resolve, reject) => {
      if (this.isScanning) {
        reject('Scanning wallet.')
      }

      client.createAddress(
        { ignoreMaxGap: true },
        (error, address: Address) => {
          if (error) {
            logger.error(error)
            reject(error)
          }
          return resolve(address)
        },
      )
    })
  }

  public getTransactionHistory(
    limit: number = 10,
  ): Promise<WalletTransaction[]> {
    return new Promise((resolve, reject) => {
      if (this.isScanning) {
        reject('Scanning wallet.')
      }

      client.getTxHistory({ limit }, (err, history) => {
        if (err) return reject(err)

        resolve(history as WalletTransaction[])
      })
    })
  }

  public checkIfReady(): Promise<{
    isReady: boolean
    notifications: BWSNotification[]
  }> {
    return new Promise((resolve, reject) => {
      client.getNotifications(
        {},
        (errors, notifications: BWSNotification[]) => {
          if (errors) {
            return reject(errors)
          }

          const notificationsfiltered = notifications.filter(
            notification =>
              notification.type === 'ScanFinished' &&
              !this.scannedErrors.includes(notification.id),
          )

          this.scannedErrors = [
            ...this.scannedErrors,
            ...notificationsfiltered.map(n => n.id),
          ]

          if (
            notificationsfiltered.length > 0 &&
            notificationsfiltered[0] &&
            notificationsfiltered[0].data &&
            notificationsfiltered[0].data.result === 'error'
          ) {
            client.startScan({}, () => {})
            this.isScanning = true
            resolve({
              isReady: false,
              notifications: this.filterShownNotifications(notifications),
            })
            return
          }

          resolve({
            notifications: this.filterShownNotifications(notifications),
            isReady: true,
          })
          this.isScanning = false
        },
      )
    })
  }

  private filterShownNotifications(
    notifications: BWSNotification[],
  ): BWSNotification[] {
    return notifications.filter(
      notification =>
        !notification.type.includes('NewBlock') ||
        (!notification.type.includes('ScanFinished') ||
          notification.data.result !== 'error'),
    )
  }

  /**
   * Creates, proposes and broadcasts a transaction when requesting it
   * @param passphrase your personal passphrase
   * @param address the crypto address to send your address to
   * @param amount the amount of your transaction in XVG (will be multiplied to sats automatically)
   */
  sendTransaction(
    address: string,
    amount: number,
  ): Promise<OutgoingTransaction> {
    invariant(
      this.password,
      'Something wrong, you should be unlocked by now already ...',
    )

    return new Promise((resolve, reject) => {
      if (this.isScanning) {
        reject('Scanning wallet.')
      }

      this.client.createTxProposal(
        {
          outputs: [
            {
              toAddress: address,
              amount: amount * Constants.satoshiDivider,
              message: '', // not used. bullshit. we are a currency not a note transfer :)
            },
          ],
        },
        (createError, txp) => {
          if (createError) {
            return reject(createError)
          }

          // SIGN PROPOSAL
          //
          client.publishTxProposal({ txp }, (publishError, processedTxp) => {
            if (publishError) {
              return reject(publishError)
            }

            if (!publishError) {
              // BROADCAST PROPOSAL
              client.signTxProposal(
                processedTxp,
                this.password,
                (signingError, stxp) => {
                  if (signingError) {
                    return reject(signingError)
                  }

                  client.broadcastTxProposal(
                    processedTxp,
                    (boradcastErr, tx) => {
                      if (boradcastErr) {
                        return reject(boradcastErr)
                      }
                      return resolve(tx as OutgoingTransaction)
                    },
                  )
                },
              )
            }
          })
        },
      )
    })
  }

  public async rescanWallet() {
    client.getStatus((error, status) => {
      if (error && error.code && error.code.includes('NOTAUTHORIZED')) {
        client.recreateWallet(err => {
          if (err) {
            logger.error(err)
          } else {
            client.getStatus((statusErr, status) => {
              if (statusErr) {
                logger.error(statusErr)
              } else {
                logger.info('Status: ', status)
                client.startScan({}, () => {
                  logger.info('Scan triggered')
                  this.isScanning = true
                })
              }
            })
          }
        })
      } else {
        /*client.startScan({}, () => {
          logger.info('Scan triggered')
          this.isScanning = true
        })*/
      }
    })
  }

  public getStatus(): Promise<any> {
    return new Promise((resolve, reject) => {
      if (this.isScanning) {
        reject('Scanning wallet.')
      }
      client.getStatus((error, status) => {
        if (error) {
          return reject(error)
        }

        return resolve(status)
      })
    })
  }

  /**
   * Restores a wallet with a given mnmoic
   * Basic restoring of keys and additional information
   *
   * No transactions or Balances or used Addresses
   * Those have to be synced afterwards manually.
   */
  public async restoreWalletFromMnemoic(
    mnemonic: string,
    passphrase: string,
  ): Promise<boolean> {
    invariant(!!mnemonic, 'Mnemonic is required and can´t be emtpy')
    invariant(!!passphrase, 'Passphrase is required and can´t be emtpy')
    // check that we aren't mindlessly overwriting wallets
    invariant(
      !this.isWalletAlreadyExistent(),
      'Existing wallet can`t be overriden!',
    )

    try {
      await client.seedFromMnemonic(mnemonic, {
        passphrase,
        account: settings.DEFAULT_ACCOUNT,
        network: settings.NETWORK,
      })
      return true
    } catch (e) {
      throw new Error(
        'Your mnemonic/passphrase wasn`t recognized correctly.\n' +
          'Either it is in a wrong language or you have left our some words.\n' +
          'Please also check your passphrase.',
      )
    }
  }

  public getMnemonic(): string {
    return this.client.credentials.mnemonic
  }
}

export default new VergeLightClient(client)
