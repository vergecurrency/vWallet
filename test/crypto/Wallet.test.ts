import { VergeLightClient } from '../../src/crypto/Wallet'
import { bwsClientMock } from '../mocks/BWSClientMock'
import VergeCacheStore from '../../src/stores/VergeCacheStore'

describe('[Component] Wallet management system', () => {
  beforeEach(() => {
    VergeCacheStore.clean()
  })

  test('should fail if passphrase is empty', () => {
    const client = new VergeLightClient({})
    expect(() => client.createNewWallet('')).toThrowError(
      'No Empty passphrase allowed',
    )
  })

  test('Wallet creation should work and return a valid mnemonic', async () => {
    const client = new VergeLightClient(bwsClientMock())

    expect(client.isWalletReady()).toBe(false)
    expect(client.isWalletLocked()).toBe(false)
    expect(client.isWalletAlreadyExistent()).toBe(false)

    const data = await client.createNewWallet('mypass')

    expect(data).toBeDefined()
    expect(data.mnemonic).toBeDefined()
    expect(typeof data.mnemonic).toBe('string')
    expect(data.mnemonic.split(' ').length).toBe(12)
  })

  test('should fail if passphrase is empty', async () => {
    const client = new VergeLightClient(bwsClientMock())

    expect(client.isWalletReady()).toBe(false)
    expect(client.isWalletLocked()).toBe(false)
    expect(client.isWalletAlreadyExistent()).toBe(false)

    expect(() =>
      client.createNewWallet(undefined),
    ).toThrowErrorMatchingSnapshot()

    expect(client.isWalletReady()).toBe(false)
    expect(client.isWalletLocked()).toBe(false)
    expect(client.isWalletAlreadyExistent()).toBe(false)
  })

  test('should stop creating a new wallet, if one already exists', async () => {
    const client = new VergeLightClient(bwsClientMock())
    await client.createNewWallet('mypass')

    expect(client.isWalletReady()).toBe(true)
    expect(client.isWalletLocked()).toBe(false)
    expect(client.isWalletAlreadyExistent()).toBe(true)

    expect(() =>
      client.createNewWallet('newwalletpassword'),
    ).toThrowErrorMatchingSnapshot()
  })
})
