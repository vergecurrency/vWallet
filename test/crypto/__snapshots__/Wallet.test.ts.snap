// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component] Wallet management system should fail if passphrase is empty 1`] = `"No Empty passphrase allowed"`;

exports[`[Component] Wallet management system should stop creating a new wallet, if one already exists 1`] = `"Existing wallet can\`t be overriden!"`;
